Use ctest to Run SystemTests on the CI

cailafinn opened the pull request:
### Description of work
Following on from #40296 - now is the step where we replace the old system test runner with ctest in the CI scripts. 
This PR is also where we'll fix any failing tests.

Closes #34850 


### To test:
1. Check all tests run on the CI.
2. Check all tests pass on the CI.

Build from Branch: 
[![Build Status](https://builds.mantidproject.org/buildStatus/icon?job=build_branch&build=1529)](https://builds.mantidproject.org/job/build_branch/1529/)



<!--  GATEKEEPER: When squashing, remove the section from HERE...  -->
---

### Reviewer

**Your comments will be used as part of the gatekeeper process.** Comment clearly on what you have checked and tested during your review. Provide an audit trail for any changes requested.

As per the [review guidelines](http://developer.mantidproject.org/ReviewingAPullRequest.html):

- Is the code of an acceptable quality? ([Code standards](http://developer.mantidproject.org/Standards/)/[GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html))
- Has a thorough functional test been performed? Do the changes handle unexpected input/situations?
- Are appropriately scoped unit and/or system tests provided?
- Do the release notes conform to the [guidelines](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html) and describe the changes appropriately?
- Has the relevant (user and developer) documentation been added/updated?
- If the PR author isn’t in the `mantid-developers` or `mantid-contributors` teams, add a review comment `rerun ci` to authorize/rerun the CI

### Gatekeeper

As per the [gatekeeping guidelines](https://developer.mantidproject.org/Gatekeeping.html):

- Has a thorough first line review been conducted, including functional testing?
- At a high-level, is the code quality sufficient?
- Are the base, milestone and labels correct?
<!--  GATEKEEPER: ...To HERE  -->

cailafinn committed:
Use ctest for systemtests in CI.

RE #34850

cailafinn committed:
Fix odd pathing on SANS tests

RE #34850

cailafinn committed:
Fix DrILLProcessTest

RE #34850

Co-authored-by: Remi Perenon <perenon@users.noreply.github.com>

cailafinn committed:
Fix incorrect initalisation of some tests

RE #34850

cailafinn committed:
Fix SaveNexusTest

RE #34850

cailafinn committed:
Try not extending sys.path all the time

RE #34850

cailafinn committed:
Avoid adding to the  data search dirs too much

It seems that the setup step can end up being run multiple times
between each cleanup, resulting in the size of the dict doubling
each time and taking exponentially longer to append to.

RE #34850

cailafinn committed:
Clean the reduction singleton between test runs

The ctest implementation results in the ReductionSingleton being
carried over between tests within a single test file. The singleton
has a Clean() function that needs to be called at the start of each
test - and probably should have always been anyway.

RE #34850

cailafinn committed:
Add some more data search dirs

RE #34850

cailafinn committed:
Add missing test files to systest data

RE #34850

jhaigh0 commented on buildconfig/Jenkins/Conda/run-tests:
```diff
@@ -82,9 +82,6 @@ if [[ $OSTYPE == "msys"* ]]; then
     export PYTHONHOME=$CONDA_PREFIX
-    SYSTEM_TEST_EXECUTABLE=$WORKSPACE/build/systemtest.bat
-else
-    SYSTEM_TEST_EXECUTABLE=$WORKSPACE/build/systemtest
```
could the old scripts be deleted?

cailafinn committed:
Split up DirectILLAutoProcessTest

RE #34850

jhaigh0 commented:
Looks like the test failures in the build_branch job are normally time-outs (lasting exactly 20mins). Tom suggests setting the time-out limit for system tests to be higher. It also seems like only 292 system tests are being run, I thought there were more (~700?).

cailafinn committed:
Increase timeout to 30 mins

The systemtests are longer now. Since they're run on a per-file basis
rather than a per-class basis some of the longer-running tests have
more of a tendancy to time out before completing.

RE #34850

cailafinn committed:
Skip DrillProcessTest on Windows

ILL don't run anything on windows, and it's causing problems.

RE #34850

cailafinn committed:
Output the system test logs to a file

RE #34850

cailafinn committed:
Remove now-reduntant check for junit xml

This is now covered by the previous line's check for xunit ctest failure
reports.

RE #34850

jhaigh0 reviewed (approved):
With the changes to the test xml archiving I think this is good to go

thomashampson reviewed (changes requested):
Jenkins is only parsing the system test results and not the unit test results. I'm going to re-run the job to see if it's been fixed by the github actions PR.

cailafinn committed:
Output systest to junit

RE #34850

cailafinn committed:
Update SaveNexusTest ignore list

RE #34850

thomashampson reviewed (commented):
rerun ci

thomashampson commented on Testing/SystemTests/tests/framework/SaveNexusTest.py:
```diff
@@ -52,12 +57,14 @@
-class LoadAndSaveLotsOfInstruments(object):
+class LoadAndSaveLotsOfInstruments(systemtesting.MantidSystemTest):
     def __getDataFileList__(self):
         # get a list of directories to look in
```
```suggestion
        # Get a list of directories to look in
```

thomashampson committed:
trivial comment change to force ci to rerun

cailafinn committed:
Fix failures found by precommit

RE #34850

cailafinn committed:
Remove now unused system test file references

RE #34850

thomashampson committed:
wip - Don't write config to file

thomashampson committed:
Disable most unit tests to speed up feedback time

thomashampson commented:
closing while we work on this within a fork
